File: src/net/Defracted/AdminUtils/listeners/ConnectionListener.java
package net.Defracted.AdminUtils.listeners;

import net.Defracted.AdminUtils.Main;
import net.Defracted.AdminUtils.util.Formatters;
import org.bukkit.BanEntry;
import org.bukkit.BanList;
import org.bukkit.Bukkit;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerPreLoginEvent;

import java.net.InetAddress;
import java.util.Date;

public class ConnectionListener implements Listener {
    private Main plugin;

    public ConnectionListener(Main instance) {
        plugin = instance;
    }

    @EventHandler(priority = EventPriority.HIGH)
    public void onPlayerLogin(AsyncPlayerPreLoginEvent event) {
        if (event == null) return;

        String name = event.getName();
        InetAddress addr = event.getAddress();


        // Ban IP
        if (Bukkit.getBanList(BanList.Type.IP).isBanned(addr.getHostAddress())) {
            BanEntry entry = Bukkit.getBanList(BanList.Type.IP).getBanEntry(addr.getHostAddress());

            if (entry == null) {
                return;
            }

            String reason = entry.getReason();
            String message = Formatters.chat("&cВаш IP-адрес забанен!\n\n&7Причина: &f" + reason);
            event.disallow(AsyncPlayerPreLoginEvent.Result.KICK_BANNED, message);
            return;
        }

        // Ban Player
        if (Bukkit.getBanList(BanList.Type.NAME).isBanned(name)) {
            BanEntry entry = Bukkit.getBanList(BanList.Type.NAME).getBanEntry(name);

            if (entry == null) {
                return;
            }

            String reason = entry.getReason();
            Date expiration = entry.getExpiration();

            String punishmentDurationStr;
            if (expiration == null) {
                punishmentDurationStr = null;
            } else {
                punishmentDurationStr = expiration.toString();
            }

            if (reason != null) {
                String message;

                if (expiration == null) {
                    message = Formatters.chat("&cВы забанены навсегда!\n\n");
                } else {
                    message = Formatters.chat("&cВы временно забанены!\n\n&7Бан действует до: &f" + punishmentDurationStr + "\n");
                }

                message += Formatters.chat("&7Причина: &f" + reason + "\n\n&7Разбан: &b&n" + plugin.getConfig().getString("appeal_url"));

                event.disallow(AsyncPlayerPreLoginEvent.Result.KICK_BANNED, message);
                return;
            }
        }

    }
}

File: src/main/kotlin/net/Defracted/AdminUtils/listeners/ConnectionListener.kt
package net.Defracted.AdminUtils.listeners

import net.Defracted.AdminUtils.Main
import net.Defracted.AdminUtils.util.Formatters.chat
import org.bukkit.BanList
import org.bukkit.Bukkit
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerPreLoginEvent
import java.util.*


class ConnectionListener(private val plugin: Main) : Listener {
    @EventHandler(priority = EventPriority.HIGH)
    fun onPlayerLogin(event: AsyncPlayerPreLoginEvent?) {
        if (event == null) return
        val name = event.name
        val addr = event.address


        // Ban IP
        if (Bukkit.getBanList(BanList.Type.IP).isBanned(addr.hostAddress)) {
            val entry = Bukkit.getBanList(BanList.Type.IP).getBanEntry(addr.hostAddress) ?: return

            val reason = entry.reason
            val message = chat("&cВаш IP-адрес забанен!\n\n&7Причина: &f$reason")
            event.disallow(AsyncPlayerPreLoginEvent.Result.KICK_BANNED, message)
            return
        }

        // Ban Player
        if (Bukkit.getBanList(BanList.Type.NAME).isBanned(name)) {
            val entry = Bukkit.getBanList(BanList.Type.NAME).getBanEntry(name) ?: return
            val reason = entry.reason
            val expiration: Date? = entry.expiration
            val punishmentDurationStr: String?
            punishmentDurationStr = expiration?.toString()

            if (reason != null) {
                var message: String = if (expiration == null) {
                    chat("&cВы забанены навсегда!\n\n")
                } else {
                    chat("&cВы временно забанены!\n\n&7Бан действует до: &f$punishmentDurationStr\n")
                }

                message += chat("&7Причина: &f$reason\n&7Разбан: &b&n${plugin.config.getString("appeal_url")}")
                event.disallow(AsyncPlayerPreLoginEvent.Result.KICK_BANNED, message)
                return
            }
        }
    }
}